### PR TITLE
Face alignment and confidence.

### PR DESCRIPTION
**Pull request by matiasdelellis**
Hi @stalker314314 
It's quite simple, just to show the idea ... and see what you think..

**Done:**
 - [x] Add column confidence and landmarks,
 - [x] Add an simple update step that fill this values.
 - [x] Take the confidence value in consideration to clustering or ignore the face.

**TODO:**
 - [x] Fill with real data.
 - [x] Admin page.. Add setting similar to sensivility..
 - [ ] Fronted..??? I show them ??
 - [ ] Everything that appears in the process.. 

NOTE: I will fill in the marks even if they are useless ..Then we see what to do with them .. :sweat_smile:

### CODE REVIEW

File: lib/Db/Face.php [hunk 1]
```diff
@@ -86,6 +88,20 @@ class Face extends Entity implements JsonSerializable {
...
 
+	/**
+	 * Confidence of face detection obtained from the model
+	 *
+	 * @var float
+	 * */
+	public $confidence;
+
+	/**
+	 * 5x2D of 68x2D landmarks for this face.
```
Comment (stalker314314) [inline lib/Db/Face.php]: same here - I would not go into "implementation details" of model:)

**Comment by stalker314314** [comment]
BTW, I approved adding these columns of confidence and landmarks (I think it is great to have them, shame we didn't already have them!). One suggestion though: if you plan to use those columns, create new PR, please!

File: appinfo/database.xml [hunk 2]
```diff
@@ -235,6 +235,21 @@
 				<length>4</length>
 				<comments>Bottom border of bounding rectangle for this face</comments>
 			</field>
+			<field>
+				<name>confidence</name>
+				<type>float</type>
+				<notnull>true</notnull>
+				<length>4</length>
+				<comments>Confidence of face detection obtained from the model used</comments>
```
[the remark this answers is not in the document]
Comment (matiasdelellis) [reply appinfo/database.xml]: Still, can't find any documentation about this value ..
But ideally it is greater than 1 .. (Although very close), and decreases greatly when the image quality decreases. Depend on how the model is trained.
I have no better answer..  but in practice we will see how reliable it is ..

File: appinfo/database.xml [hunk 3]
```diff
@@ -235,6 +235,21 @@
...
+			</field>
+			<field>
+				<name>landmarks</name>
+				<!-- todo: Doctrine obviusly allows json type, but I tested with PG only, need to test with SQLite and MySQL too -->
+				<type>json_array</type>
```
[the remark this answers is not in the document]
Comment (matiasdelellis) [reply appinfo/database.xml]: Ok.
You're right .. it was just a cut and paste of the descriptor .. :sweat_smile:

File: appinfo/database.xml (hunk 2 above)
[the remark this answers is not in the document]
Comment (stalker314314) [reply appinfo/database.xml]: Oh, 💩 . I though it is [0-1]. If it is not, forget my comment. We cannot normalize it then:/

**Comment by matiasdelellis** [comment]
> BTW, I approved adding these columns of confidence and landmarks (I think it is great to have them, shame we didn't already have them!). One suggestion though: if you plan to use those columns, create new PR, please!.

OK no problem ...
Personally I thought to analyze my photos again :see_no_evil:  and see what can really be done with these values before continuing ..

**Comment by stalker314314** [comment]
It is just a suggestion. I think there is big value to have confidence, even if all experiment fails! After all, this app is about creating API for others to use face data, right?;)

**Comment by matiasdelellis** [comment]
> What is plan with 68points model? Add it as model with id=2? Or make 68 point model default (id=1) and offer 5 point model as id=2?

As planned, we should do as model id=2, but for now it doesn't worry me..
About the position of the faces, the model 68 does not improve the result, and clearly uses more memory. Therefore, it would not be my main option.

**Comment by matiasdelellis** [comment]
Hi @stalker314314 

> It is just a suggestion. I think there is big value to have confidence, even if all experiment fails! After all, this app is about creating API for others to use face data, right?;)

Well...
This little experiment, ended up being very important .. :grimacing: 

See last comment on: https://github.com/matiasdelellis/facerecognition/issues/114#issuecomment-535901466

..therefore I will fix your suggestions, and merge this complete branch..

> > One suggestion though: if you plan to use those columns, create new PR, please!
>
> OK no problem ...

I know I said this, but the result is overwhelming, and the only change in processing is this line:

https://github.com/matiasdelellis/facerecognition/pull/164/files#diff-324fc525c91828dc868593c5154275a8R229-R232

Any problem with this?

**Comment by stalker314314** [comment]
One think that I just thought of that is missing is: invalidation of X (where X is clustering or whole pipeline) when confidence is changed (similar how we reset everything when sensitivity is changed). X depends on how you end up doing it.

If you ignore faces with low confidence in clustering logic, then you need to put all persons as invalid (basically force next clustering logic).

If you ignore faces with low confidence immediately when detected (I prefer that, as explained in issue #114 ), then whole pipeline needs to be invalidated (there could be some optimization later, if you go with low confidence to higher, you could just delete faces in between and invalidate clustering logic...)

**Comment by matiasdelellis** [comment]
> One think that I just thought of that is missing is: invalidation of X (where X is clustering or whole pipeline) when confidence is changed (similar how we reset everything when sensitivity is changed). X depends on how you end up doing it.
> 
> If you ignore faces with low confidence in clustering logic, then you need to put all persons as invalid (basically force next clustering logic).
> 
> If you ignore faces with low confidence immediately when detected (I prefer that, as explained in issue #114 ), then whole pipeline needs to be invalidated (there could be some optimization later, if you go with low confidence to higher, you could just delete faces in between and invalidate clustering logic...)

Completely discards the information of the face implies a reprocessing the images to obtain new faces.
..and we don't even have the logic to do this correctly.

...changing the value of the confidence, does not need to invalidate any face, or person. The current information is valid. 

> similar how we reset everything when sensitivity is changed

As for reprocessing, it now works exactly the same. When changing the confidence, just mark that all users must regroup, but the current information is not invalidated.

Try to make this PR as less invasive as possible. If you use a low value, the operation is exactly the same as until now.

**Comment by stalker314314** [comment]
Agree to make it least invasive, just go ahead, we can debate what to do with faces < confidence in some other issue/PR:D

**Comment by matiasdelellis** [comment]
:sweat: 
These days I have little time again... So I merge it and move on other issues.. :wink: 

p.s: Reading my answers, maybe it seems aggressive. Please, if you felt that way, in any way it was with intention. :disappointed:

**Comment by stalker314314** [comment]
Ha ha, don't worry, I am not easily offended, you will have to try REALLY hard for that:)

**Pull request merged**